### PR TITLE
[3.x] Change: Allow architecture specific dependancies

### PR DESCRIPTION
**Pull request by leezer3**
### Purpose of this PR

TLDR:
I'm planning on shipping a 64-build at some not too distant stage in the future, but for legacy reasons I still want to maintain 32-bit compatability also.

I currently ship the 32-bit version of OpenALSoft, but this obviously doesn't work on 64-bit.

This PR is dead simple, and allows architecture specific PInvokes to SDL and OpenALSoft, by using the common convention of adding **x86** and **x64** to the DLL search directory list when the tookit is initialised.
Only potential flaw I can see is if OpenAL is called before attempting to init the toolkit itself (is this even possible? I haven't tried....)

### Testing status

* Simple adding of a alternate path to the PInvoke search directories using native Windows function. Works fine.

### Comments

n/a

### CODE REVIEW

**Comment by NogginBops** [comment]
This will only work on windows though? 
Mono doesn't implement functions from `kernel32.dll`?

**Comment by leezer3** [comment]
Correct in part.

Mono just uses the .config file to map the Windows library names to .so equivilants, which IIRC works whatever.

Might need an platform switch to disable the kernel32 call though, will have to test Linux.

**Comment by leezer3** [comment]
You're entirely right there, Linux needed a platform check.

Would have hoped that Mono might handle it internally, but nope.

**Comment by varon** [comment]
muh ci

File: src/OpenTK/Toolkit.cs [hunk 1]
```diff
@@ -129,6 +132,18 @@ public static Toolkit Init(ToolkitOptions options)
                     initialized = true;
                     Configuration.Init(options);
                     Options = options;
+                    if (Environment.OSVersion.Platform == PlatformID.Win32S | Environment.OSVersion.Platform == PlatformID.Win32Windows | Environment.OSVersion.Platform == PlatformID.Win32NT)
```
Comment (NogginBops) [inline src/OpenTK/Toolkit.cs:135]: This this not use the logical or operator `||` instead?

**Review by NogginBops — CHANGES_REQUESTED** [review]
This seems like a reasonable change to make.
What I would want from this PR though is some more documentation on how this is supposed to be used. 
As at the moment it's unclear where 32 bit vs 64 bit dependencies are supposed to be placed and the codebase doesn't have any comments or documents to explain this.

So I would really appreciate it if this feature could get a written down explanation of how to use it (basically just describe the different folders, where they should be placed and what they do).

File: src/OpenTK/Toolkit.cs (hunk 1 above)
Comment (leezer3) [reply src/OpenTK/Toolkit.cs:135]: Can change if you want, especially as it's not in a performance critical area, although this one would be insignificantly faster, & I've done it this way out of habit :P

**Comment by leezer3** [comment]
Where did you want documentation put?

I can write a simple dependancies.md or something, or append it to the readme if you'd rather.
At the minute, any dependancies are probed for as per this MS document:
https://docs.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order
(We can probably safely assume that any normal end user will have safe mode searching on)

File: src/OpenTK/Toolkit.cs (hunk 1 above)
Comment (NogginBops) [reply src/OpenTK/Toolkit.cs:135]: It's not a performance concern it's a readability and correctness term. The precidence between `|` and `==` is not immediately obvious sand generally it's expected that is you are working with booleans you use logical operators instead of bitwise operators.

**Comment by NogginBops** [comment]
If you could it would be great if you could write a bit about the feature as a comment in the code, basically linking the Microsoft docs.
If you wanted to you could also write a little section about it in the readme, that would be great.

**Comment by leezer3** [comment]
Done.

I've broken out a full description of all the optional libraries, DLLMaps and search paths into a separate file linked from the main readme.

**Review by NogginBops — APPROVED** [review]
Absolutely fantastic documentation! 
Thanks a ton for the PR.

Adding this PR to the 3.3.2 milestone, and will merge it when the other 3.3.2 PRs are ready.

**Comment by varon** [comment]
@NogginBops  - let's get this in! We can do a release when needed down the line.

**Comment by NogginBops** [comment]
Turns out that this PR breaks integration tests. Not 100% sure what the issue is yet, but we will want to fix this issue before making a new OpenTK 3 release.

**Comment by NogginBops** [comment]
Also, the exception that is potentially being thrown needs to get a better error message so that if you got that issue you could actually figure out what is going wrong.

**Comment by leezer3** [comment]
I can't see immediately anything that could go wrong :/
It's essentially a single P/Invoke call, and *shouldn't* generate any managed exceptions except the last (and if that .

Some marginal thoughts:
https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getentryassembly?view=net-5.0
**Entry.GetEntryAssembly.Location()**
This may return null under some conditions, specifically if a managed assembly has been loaded by an unmanaged one. I'd have thought that F# wouldn't match that pattern.
Check for null here, and just return?

**SetDllDirectory**
I suppose could wrap it in a try / catch block, in case it's something funky going on with UAC on the test runner?
Has been supereceded on later Windows versions, but *should* work: https://github.com/FNA-XNA/FNA/issues/162

**Comment by NogginBops** [comment]
It's because of a null reference at line `155`, so the theory of `Assembly.GetEntryAssembly().Location` being `null` seems very likely.

**Comment by leezer3** [comment]
@NogginBops 
Have opened the following PR to make that change, wrap in try / catch, and finally, I've decided to limit the exception to debug builds only.
https://github.com/opentk/opentk/pull/1312
